In [1]:
# loading reviews.txt
review_data = open('data/reviews.txt','r')
reviews = list(map(lambda x:x[:-1],review_data.readlines()))
len(reviews)

25000

In [2]:
#loading labels.txt
label_data = open('data/labels.txt','r')
labels = list(map(lambda x:x[:-1],label_data.readlines()))
len(labels)

25000

In [3]:
reviews[5]

'this film lacked something i couldn  t put my finger on at first charisma on the part of the leading actress . this inevitably translated to lack of chemistry when she shared the screen with her leading man . even the romantic scenes came across as being merely the actors at play . it could very well have been the director who miscalculated what he needed from the actors . i just don  t know .  br    br   but could it have been the screenplay  just exactly who was the chef in love with  he seemed more enamored of his culinary skills and restaurant  and ultimately of himself and his youthful exploits  than of anybody or anything else . he never convinced me he was in love with the princess .  br    br   i was disappointed in this movie . but  don  t forget it was nominated for an oscar  so judge for yourself .  '

In [4]:
labels[5]

'negative'

In [5]:
def mapping_review_and_label(i):
    print(labels[i] + "\t:\t" + reviews[i][:100] + "...")

In [6]:
print("LABELS \t\t:\t\t\t REVIEWS\n")
mapping_review_and_label(1200)
mapping_review_and_label(2531)
mapping_review_and_label(3506)
mapping_review_and_label(4509)
mapping_review_and_label(15000)

LABELS 		:			 REVIEWS

positive	:	i can only agree with taximeter that this is a fantastic film and should be seen by a wide audience ...
negative	:	there isn  t much to say about this film  it is horrible .  br    br   the acting and dialog are way...
positive	:	this movie is a masterpiece a story of a young woman during the war  and it really happen  not exact...
negative	:	generally  i  ve found that if you don  t hear about a movie prior to seeing it on dvd  there  s pro...
positive	:	this documentary  now available free on video . google . com  is a fantastic demonstration of the po...


In [7]:
from collections import Counter
import numpy as np

In [8]:
positive_wordcounts = Counter()# counter object for words from positive reviews
negative_wordcounts = Counter()# counter object for words from negative reviews
total_wordcounts = Counter()# # counter object for total words from all reviews

# loop for running through all the reviews and adding the count in respective counter object(positive,negative,totall)
for i in range(len(reviews)):
    if(labels[i] == 'positive'):
        for word in reviews[i].split(" "):
            positive_wordcounts[word] += 1
            total_wordcounts[word] += 1
    else:
        for word in reviews[i].split(" "):
            negative_wordcounts[word] += 1
            total_wordcounts[word] += 1

In [9]:
# Examining the counts of the most common words in positive reviews
positive_wordcounts.most_common()

[('', 550468),
 ('the', 173324),
 ('.', 159654),
 ('and', 89722),
 ('a', 83688),
 ('of', 76855),
 ('to', 66746),
 ('is', 57245),
 ('in', 50215),
 ('br', 49235),
 ('it', 48025),
 ('i', 40743),
 ('that', 35630),
 ('this', 35080),
 ('s', 33815),
 ('as', 26308),
 ('with', 23247),
 ('for', 22416),
 ('was', 21917),
 ('film', 20937),
 ('but', 20822),
 ('movie', 19074),
 ('his', 17227),
 ('on', 17008),
 ('you', 16681),
 ('he', 16282),
 ('are', 14807),
 ('not', 14272),
 ('t', 13720),
 ('one', 13655),
 ('have', 12587),
 ('be', 12416),
 ('by', 11997),
 ('all', 11942),
 ('who', 11464),
 ('an', 11294),
 ('at', 11234),
 ('from', 10767),
 ('her', 10474),
 ('they', 9895),
 ('has', 9186),
 ('so', 9154),
 ('like', 9038),
 ('about', 8313),
 ('very', 8305),
 ('out', 8134),
 ('there', 8057),
 ('she', 7779),
 ('what', 7737),
 ('or', 7732),
 ('good', 7720),
 ('more', 7521),
 ('when', 7456),
 ('some', 7441),
 ('if', 7285),
 ('just', 7152),
 ('can', 7001),
 ('story', 6780),
 ('time', 6515),
 ('my', 6488),
 ('g

In [10]:
# Examining the counts of the most common words in negative reviews
negative_wordcounts.most_common()

[('', 561462),
 ('.', 167538),
 ('the', 163389),
 ('a', 79321),
 ('and', 74385),
 ('of', 69009),
 ('to', 68974),
 ('br', 52637),
 ('is', 50083),
 ('it', 48327),
 ('i', 46880),
 ('in', 43753),
 ('this', 40920),
 ('that', 37615),
 ('s', 31546),
 ('was', 26291),
 ('movie', 24965),
 ('for', 21927),
 ('but', 21781),
 ('with', 20878),
 ('as', 20625),
 ('t', 20361),
 ('film', 19218),
 ('you', 17549),
 ('on', 17192),
 ('not', 16354),
 ('have', 15144),
 ('are', 14623),
 ('be', 14541),
 ('he', 13856),
 ('one', 13134),
 ('they', 13011),
 ('at', 12279),
 ('his', 12147),
 ('all', 12036),
 ('so', 11463),
 ('like', 11238),
 ('there', 10775),
 ('just', 10619),
 ('by', 10549),
 ('or', 10272),
 ('an', 10266),
 ('who', 9969),
 ('from', 9731),
 ('if', 9518),
 ('about', 9061),
 ('out', 8979),
 ('what', 8422),
 ('some', 8306),
 ('no', 8143),
 ('her', 7947),
 ('even', 7687),
 ('can', 7653),
 ('has', 7604),
 ('good', 7423),
 ('bad', 7401),
 ('would', 7036),
 ('up', 6970),
 ('only', 6781),
 ('more', 6730),
 ('

As we can see from the above result that words such as it,this,that,the, etc. are frequent in both type of reviews and has no sentimental meaning.
We need to check which words are more common in positive reviews than negative reviews and vice versa
For this we will calculate positive to negative ratio.

In [11]:
pos_to_neg_ratios = Counter()

for word,count in list(total_wordcounts.most_common()):
    if(count > 100):# only consider words that occurs more than 100 times in reviews
        pos_to_neg = positive_wordcounts[word] / float(negative_wordcounts[word]+1) # +1 for avoiding divide by zero error if word is only in positive reviews
        pos_to_neg_ratios[word] = pos_to_neg

In [12]:
print("Pos-to-neg ratio for 'this' = {}".format(pos_to_neg_ratios["this"]))
print("Pos-to-neg ratio for 'wonderful' = {}".format(pos_to_neg_ratios["wonderful"]))
print("Pos-to-neg ratio for 'terrible' = {}".format(pos_to_neg_ratios["terrible"]))

Pos-to-neg ratio for 'this' = 0.8572615527479778
Pos-to-neg ratio for 'wonderful' = 4.780487804878049
Pos-to-neg ratio for 'terrible' = 0.17744252873563218


From the above results, it seems 1 is considered as neutral, less than 1 negative, greater than 1 positive. Lets bring neutrality to 0 using log

In [13]:
for word,ratio in pos_to_neg_ratios.most_common():
    if(ratio > 1):
        pos_to_neg_ratios[word] = np.log(ratio)
    else:
        pos_to_neg_ratios[word] = -np.log((1 / (ratio+0.01)))

In [14]:
print("Pos-to-neg ratio for 'this' = {}".format(pos_to_neg_ratios["this"]))
print("Pos-to-neg ratio for 'wonderful' = {}".format(pos_to_neg_ratios["wonderful"]))
print("Pos-to-neg ratio for 'terrible' = {}".format(pos_to_neg_ratios["terrible"]))

Pos-to-neg ratio for 'this' = -0.1424146720933266
Pos-to-neg ratio for 'wonderful' = 1.5645425925262093
Pos-to-neg ratio for 'terrible' = -1.6742829939664696


In [15]:
# words most frequently seen in a review with a "positive" label
pos_to_neg_ratios.most_common()

[('edie', 4.6913478822291435),
 ('paulie', 4.07753744390572),
 ('felix', 3.152736022363656),
 ('polanski', 2.8233610476132043),
 ('matthau', 2.80672172860924),
 ('victoria', 2.681021528714291),
 ('mildred', 2.6026896854443837),
 ('gandhi', 2.538973871058276),
 ('flawless', 2.451005098112319),
 ('superbly', 2.26002547857525),
 ('perfection', 2.159484249353372),
 ('astaire', 2.1400661634962708),
 ('captures', 2.038619547159581),
 ('voight', 2.030170492673053),
 ('wonderfully', 2.0218960560332353),
 ('powell', 1.978345424808467),
 ('brosnan', 1.9547990964725592),
 ('lily', 1.9203768470501485),
 ('bakshi', 1.9029851043382795),
 ('lincoln', 1.9014583864844796),
 ('refreshing', 1.8551812956655511),
 ('breathtaking', 1.8481124057791867),
 ('bourne', 1.8478489358790986),
 ('lemmon', 1.8458266904983307),
 ('delightful', 1.8002701588959635),
 ('flynn', 1.7996646487351682),
 ('andrews', 1.7764919970972666),
 ('homer', 1.7692866133759964),
 ('beautifully', 1.7626953362841438),
 ('soccer', 1.757857

## Creating Input and output data

In [16]:
reviews_vocab = set(total_wordcounts.keys())
reviews_vocab_size = len(reviews_vocab)
print(reviews_vocab_size)

74074


In [17]:
print(reviews_vocab)

{'', 'pointlessly', 'colorado', 'pratt', 'spades', 'ricki', 'feistyness', 'homogeneous', 'khmer', 'encumbered', 'gwenneth', 'unedifying', 'misdemeanor', 'inconsistency', 'arabian', 'unapologetically', 'dissaude', 'haiku', 'petri', 'xk', 'scintillating', 'kneed', 'undoubted', 'bobbins', 'ringu', 'malevolent', 'playboys', 'gangfights', 'remus', 'bathos', 'kevloun', 'cederic', 'hardie', 'eater', 'households', 'haste', 'embraced', 'entry', 'unspeakable', 'solve', 'azuma', 'portugeuse', 'sochenge', 'bonaparte', 'doppelgang', 'haneke', 'crewman', 'catalunya', 'sitra', 'dollhouse', 'cogan', 'kibbee', 'transitory', 'bianca', 'sugar', 'hastening', 'handy', 'servers', 'bumped', 'witney', 'accidence', 'monstervision', 'fei', 'whereupon', 'hallgren', 'huggie', 'principal', 'candidates', 'parenting', 'underestimates', 'seeming', 'tied', 'spinal', 'fanned', 'fiftieth', 'morpheus', 'resource', 'pliant', 'beyondreturn', 'paxinou', 'keats', 'propagandistic', 'moko', 'frazzlingly', 'hercules', 'currie',

Creating a numpy array called input_layer and initialize it to all zeros. input_layer is created as a 2-dimensional matrix with 1 row and reviews_vocab_size columns (1 x 74074).

In [18]:
import numpy as np

input_layer = np.zeros((1,reviews_vocab_size))
input_layer

array([[0., 0., 0., ..., 0., 0., 0.]])

In [19]:
input_layer.shape

(1, 74074)

In [20]:
# creating a lookup table that stores the index of every word of the vocabulary
# The lookup table is just a dictionary of words in the vocabulary mapped to index positions to be used in input_layer
word2index = {}
for i,word in enumerate(reviews_vocab):
    word2index[word] = i
    
# display the map of words to indices
word2index


{'': 0,
 'pointlessly': 1,
 'colorado': 2,
 'pratt': 3,
 'spades': 4,
 'ricki': 5,
 'feistyness': 6,
 'homogeneous': 7,
 'khmer': 8,
 'encumbered': 9,
 'gwenneth': 10,
 'unedifying': 11,
 'misdemeanor': 12,
 'inconsistency': 13,
 'arabian': 14,
 'unapologetically': 15,
 'dissaude': 16,
 'haiku': 17,
 'petri': 18,
 'xk': 19,
 'scintillating': 20,
 'kneed': 21,
 'undoubted': 22,
 'bobbins': 23,
 'ringu': 24,
 'malevolent': 25,
 'playboys': 26,
 'gangfights': 27,
 'remus': 28,
 'bathos': 29,
 'kevloun': 30,
 'cederic': 31,
 'hardie': 32,
 'eater': 33,
 'households': 34,
 'haste': 35,
 'embraced': 36,
 'entry': 37,
 'unspeakable': 38,
 'solve': 39,
 'azuma': 40,
 'portugeuse': 41,
 'sochenge': 42,
 'bonaparte': 43,
 'doppelgang': 44,
 'haneke': 45,
 'crewman': 46,
 'catalunya': 47,
 'sitra': 48,
 'dollhouse': 49,
 'cogan': 50,
 'kibbee': 51,
 'transitory': 52,
 'bianca': 53,
 'sugar': 54,
 'hastening': 55,
 'handy': 56,
 'servers': 57,
 'bumped': 58,
 'witney': 59,
 'accidence': 60,
 'mons

In [21]:
# function to modify the global input_layer to represent the vector form of review.
# The element at a given index of input_layer should represent how many times the given word occurs in the review.
    
def reset_input_layer(review):
     
    global input_layer
    
    # clear out previous state of input_layer and reset the layer to be all 0's
    input_layer *= 0
    
    # count how many times each word is used in the given review and store the results in input_layer 
    for word in review.split(" "):
        input_layer[0][word2index[word]] += 1

In [22]:
reset_input_layer(reviews[0])
input_layer

array([[18.,  0.,  0., ...,  0.,  0.,  0.]])

In [23]:
# converting labels to numeric values positive=1,negative=0
def label_to_number(label):
    if(label == 'positive'):
        return 1
    else:
        return 0

In [24]:
labels[1]

'negative'

In [25]:
label_to_number(labels[1])

0

In [26]:
import numpy as np
import copy
import sys

In [27]:
def keySwitch(M,c,l):
    c_star = getBitVector(c,l)
    return M.dot(c_star)

def getRandomMatrix(row,col,bound):
    A = np.zeros((row,col))
    for i in range(row):
        for j in range(col):
            A[i][j] = np.random.randint(bound)
    return A

def getBitMatrix(S,l):
    S_star = list()
    for i in range(l):
        S_star.append(S*2**(l-i-1))
    S_star = np.array(S_star).transpose(1,2,0).reshape(len(S),len(S[0])*l)
    return S_star

def getSecretKey(T):
    assert(T.ndim == 2)
    I = np.eye(len(T)) # num rows
    return hCat(I,T)

def hCat(A,B):
    return np.concatenate((A,B),1)

def vCat(A,B):
    return np.concatenate((A,B),0)

def keySwitchMatrix(S, T,l):
    S_star = getBitMatrix(S,l)
    A = getRandomMatrix(T.shape[1],S_star.shape[1], aBound)
    E = getRandomMatrix(S_star.shape[0], S_star.shape[1], eBound)
    return vCat(S_star + E - T.dot(A), A)

def encrypt(T, x,w,l):
    return keySwitch(keySwitchMatrix(np.eye(len(x)),T,l), w * x,l)

def addVectors(c1, c2):
    return c1 + c2

def linearTransform(M, c, l):
    return M.dot(getBitVector(c, l)).astype('int64')

def linearTransformClient(G, S, T, l):
    return keySwitchMatrix(G.dot(S), T,l)

def vectorize(M):
    ans = np.zeros((len(M) * len(M[0]),1))
    for i in range(len(M)):
        for j in range(len(M[0])):
            ans[i * len(M[0]) + j][0] = M[i][j]
    return ans

def decrypt(S, c,w):
    Sc = S.dot(c)
    return (Sc / w).astype('float').round().astype('int')

def innerProdClient(T,l):
    S = getSecretKey(T)
    tvsts = vectorize(S.T.dot(S)).T
    mvsts = copyRows(tvsts, len(T))
    return keySwitchMatrix(mvsts,T,l)

def copyRows(row, numrows):
    ans = np.zeros((numrows, len(row[0])))
    for i in range(len(ans)):
        for j in range(len(ans[0])):
            ans[i][j] = row[0][j]
            
    return ans

def innerProd(c1, c2, M,l):
    
    cc1 = np.zeros((len(c1),1))
    for i in range(len(c1)):
        cc1[i][0] = c1[i]
    
    cc2 = np.zeros((1, len(c2)))
    for i in range(len(c2)):
        cc2[0][i] = c2[i]
        
    cc = vectorize(cc1.dot(cc2))
    
    bv = getBitVector((cc / w).round().astype('int64'),l)
    
    return M.dot(bv)

def one_way_encrypt_vector(vector,scaling_factor = 1000):
    padded_vector = np.random.rand(len(vector)+1)
    padded_vector[0:len(vector)] = vector
    
    vec_len = len(padded_vector)
    
    M_temp = (M_keys[vec_len-2].T*padded_vector*scaling_factor / (vec_len-1)).T
    e_vector = innerProd(c_ones[vec_len-2],c_ones[vec_len-2],M_temp,l)
    return e_vector.astype('int64')

def load_linear_transformation(syn0_text,scaling_factor = 1000):
    syn0_text *= scaling_factor
    return linearTransformClient(syn0_text.T,getSecretKey(T),T,l)

def s_decrypt(vec):
    return decrypt(getSecretKey(T_keys[len(vec)-2]),vec,w)

def add_vectors(x,y,scaling_factor = 10000):
    return x + y

def transpose(syn1):

    rows = len(syn1)
    cols = len(syn1[0]) - 1
    
    max_rc = max(rows,cols)
    
    syn1_c = list()
    for i in range(len(syn1)):
        tmp = np.zeros(max_rc+1)
        tmp[:len(syn1[i])] = syn1[i]
        syn1_c.append(tmp)
    
    syn1_c_transposed = list()
    
    for row_i in range(cols):
        syn1t_column = innerProd(syn1_c[0],v_onehot[max_rc-1][row_i],M_onehot[max_rc-1][0],l) / scaling_factor
        for col_i in range(rows-1):
            syn1t_column += innerProd(syn1_c[col_i+1],v_onehot[max_rc-1][row_i],M_onehot[max_rc-1][col_i+1],l) / scaling_factor

        syn1_c_transposed.append(syn1t_column[0:rows+1])
    
    return syn1_c_transposed

def int2bin(x):
    s = list()
    mod = 2
    while(x > 0):
        s.append(int(x % 2))
        x = int(x / 2)
    return np.array(list(reversed(s))).astype('int64')


def getBitVector(c,l):
    m = len(c)
    c_star = np.zeros(l * m,dtype='int64')
    for i in range(m):
        local_c = int(c[i])
        if(local_c < 0):
            local_c = -local_c
        b = int2bin(local_c)
        if(c[i] < 0):
            b *= -1
        if(c[i] == 0):
            b *= 0
#         try:
        c_star[(i * l) + (l-len(b)): (i+1) * l] += b
#         except:
#             print(len(b))
#             print(i)
#             print(len(c_star[(i * l) + (l-len(b)): (i+1) * l]))
    return c_star


In [28]:
# HAPPENS ON SECURE SERVER

l = 100
w = 2 ** 25

aBound = 10
tBound = 10
eBound = 10

max_dim = 16

scaling_factor = 1000

# keys
T_keys = list()
for i in range(max_dim):
    T_keys.append(np.random.rand(i+1,1))

# one way encryption transformation
M_keys = list()
for i in range(max_dim):
    M_keys.append(innerProdClient(T_keys[i],l))

M_onehot = list()
for h in range(max_dim):
    i = h+1
    buffered_eyes = list()
    for row in np.eye(i+1):
        buffer = np.ones(i+1)
        buffer[0:i+1] = row
        buffered_eyes.append((M_keys[i-1].T * buffer).T)
    M_onehot.append(buffered_eyes)
    
c_ones = list()
for i in range(max_dim):
    c_ones.append(encrypt(T_keys[i],np.ones(i+1), w, l).astype('int'))
    
v_onehot = list()
onehot = list()
for i in range(max_dim):
    eyes = list()
    eyes_txt = list()
    for eye in np.eye(i+1):
        eyes_txt.append(eye)
        eyes.append(one_way_encrypt_vector(eye,scaling_factor))
    v_onehot.append(eyes)
    onehot.append(eyes_txt)

H_sigmoid_txt = np.zeros((5,5))

H_sigmoid_txt[0][0] = 0.5
H_sigmoid_txt[0][1] = 0.25
H_sigmoid_txt[0][2] = -1/48.0
H_sigmoid_txt[0][3] = 1/480.0
H_sigmoid_txt[0][4] = -17/80640.0

H_sigmoid = list()
for row in H_sigmoid_txt:
    H_sigmoid.append(one_way_encrypt_vector(row))


In [29]:
def sigmoid(layer_2_c):
    out_rows = list()
    for position in range(len(layer_2_c)-1):

        M_position = M_onehot[len(layer_2_c)-2][0]

        layer_2_index_c = innerProd(layer_2_c,v_onehot[len(layer_2_c)-2][position],M_position,l) / scaling_factor

        x = layer_2_index_c
        x2 = innerProd(x,x,M_position,l) / scaling_factor
        x3 = innerProd(x,x2,M_position,l) / scaling_factor
        x5 = innerProd(x3,x2,M_position,l) / scaling_factor
        x7 = innerProd(x5,x2,M_position,l) / scaling_factor

        xs = copy.deepcopy(v_onehot[5][0])
        xs[1] = x[0]
        xs[2] = x2[0]
        xs[3] = x3[0]
        xs[4] = x5[0]
        xs[5] = x7[0]

        out = mat_mul_forward(xs,H_sigmoid[0:1],scaling_factor)
        out_rows.append(out)
    return transpose(out_rows)[0]

def load_linear_transformation(syn0_text,scaling_factor = 1000):
    syn0_text *= scaling_factor
    return linearTransformClient(syn0_text.T,getSecretKey(T_keys[len(syn0_text)-1]),T_keys[len(syn0_text)-1],l)

def outer_product(x,y):
    flip = False
    if(len(x) < len(y)):
        flip = True
        tmp = x
        x = y
        y = tmp
        
    y_matrix = list()

    for i in range(len(x)-1):
        y_matrix.append(y)

    y_matrix_transpose = transpose(y_matrix)

    outer_result = list()
    for i in range(len(x)-1):
        outer_result.append(mat_mul_forward(x * onehot[len(x)-1][i],y_matrix_transpose,scaling_factor))
    
    if(flip):
        return transpose(outer_result)
    
    return outer_result

def mat_mul_forward(layer_1,syn1,scaling_factor):
    
    input_dim = len(layer_1)
    output_dim = len(syn1)

    buff = np.zeros(max(output_dim+1,input_dim+1))
    buff[0:len(layer_1)] = layer_1
    layer_1_c = buff
    
    syn1_c = list()
    for i in range(len(syn1)):
        buff = np.zeros(max(output_dim+1,input_dim+1))
        buff[0:len(syn1[i])] = syn1[i]
        syn1_c.append(buff)
    
    layer_2 = innerProd(syn1_c[0],layer_1_c,M_onehot[len(layer_1_c) - 2][0],l) / float(scaling_factor)
    for i in range(len(syn1)-1):
        layer_2 += innerProd(syn1_c[i+1],layer_1_c,M_onehot[len(layer_1_c) - 2][i+1],l) / float(scaling_factor)
    return layer_2[0:output_dim+1]

def elementwise_vector_mult(x,y,scaling_factor):
    
    y =[y]
    
    one_minus_layer_1 = transpose(y)

    outer_result = list()
    for i in range(len(x)-1):
        outer_result.append(mat_mul_forward(x * onehot[len(x)-1][i],y,scaling_factor))
        
    return transpose(outer_result)[0]

In [30]:
import time
import sys
import numpy as np

# Let's tweak our network from before to model these phenomena
class SentimentNetwork:
    def __init__(self, reviews,labels,min_count = 10,polarity_cutoff = 0.1,hidden_nodes = 8, learning_rate = 0.1):
       
        np.random.seed(1234)
    
        self.pre_process_data(reviews, polarity_cutoff, min_count)
        
        self.init_network(len(self.review_vocab),hidden_nodes, 1, learning_rate)
        
        
    def pre_process_data(self,reviews, polarity_cutoff,min_count):
        
        print("Pre-processing data...")
        
        positive_counts = Counter()
        negative_counts = Counter()
        total_counts = Counter()

        for i in range(len(reviews)):
            if(labels[i] == 'POSITIVE'):
                for word in reviews[i].split(" "):
                    positive_counts[word] += 1
                    total_counts[word] += 1
            else:
                for word in reviews[i].split(" "):
                    negative_counts[word] += 1
                    total_counts[word] += 1

        pos_neg_ratios = Counter()

        for term,cnt in list(total_counts.most_common()):
            if(cnt >= 50):
                pos_neg_ratio = positive_counts[term] / float(negative_counts[term]+1)
                pos_neg_ratios[term] = pos_neg_ratio

        for word,ratio in pos_neg_ratios.most_common():
            if(ratio > 1):
                pos_neg_ratios[word] = np.log(ratio)
            else:
                pos_neg_ratios[word] = -np.log((1 / (ratio + 0.01)))
        
        review_vocab = set()
        for review in reviews:
            for word in review.split(" "):
                if(total_counts[word] > min_count):
                    if(word in pos_neg_ratios.keys()):
                        if((pos_neg_ratios[word] >= polarity_cutoff) or (pos_neg_ratios[word] <= -polarity_cutoff)):
                            review_vocab.add(word)
                    else:
                        review_vocab.add(word)
        self.review_vocab = list(review_vocab)
        
        label_vocab = set()
        for label in labels:
            label_vocab.add(label)
        
        self.label_vocab = list(label_vocab)
        
        self.review_vocab_size = len(self.review_vocab)
        self.label_vocab_size = len(self.label_vocab)
        
        self.word2index = {}
        for i, word in enumerate(self.review_vocab):
            self.word2index[word] = i
        
        self.label2index = {}
        for i, label in enumerate(self.label_vocab):
            self.label2index[label] = i
         
        
    def init_network(self, input_nodes, hidden_nodes, output_nodes, learning_rate):
        # Set number of nodes in input, hidden and output layers.
        self.input_nodes = input_nodes
        self.hidden_nodes = hidden_nodes
        self.output_nodes = output_nodes

        print("Initializing Weights...")
        self.weights_0_1_t = np.zeros((self.input_nodes,self.hidden_nodes))
    
        self.weights_1_2_t = np.random.normal(0.0, self.output_nodes**-0.5, 
                                                (self.hidden_nodes, self.output_nodes))
        
        print("Encrypting Weights...")
        self.weights_0_1 = list()
        for i,row in enumerate(self.weights_0_1_t):
            sys.stdout.write("\rEncrypting Weights from Layer 0 to Layer 1:" + str(float((i+1) * 100) / len(self.weights_0_1_t))[0:4] + "% done")
            self.weights_0_1.append(one_way_encrypt_vector(row,scaling_factor).astype('int64'))
        print("")
        
        self.weights_1_2 = list()
        for i,row in enumerate(self.weights_1_2_t):
            sys.stdout.write("\rEncrypting Weights from Layer 1 to Layer 2:" + str(float((i+1) * 100) / len(self.weights_1_2_t))[0:4] + "% done")
            self.weights_1_2.append(one_way_encrypt_vector(row,scaling_factor).astype('int64'))           
        self.weights_1_2 = transpose(self.weights_1_2)
        
        self.learning_rate = learning_rate
        
        self.layer_0 = np.zeros((1,input_nodes))
        self.layer_1 = np.zeros((1,hidden_nodes))
        
    def sigmoid(self,x):
        return 1 / (1 + np.exp(-x))
    
    
    def sigmoid_output_2_derivative(self,output):
        return output * (1 - output)
    
    def update_input_layer(self,review):

        # clear out previous state, reset the layer to be all 0s
        self.layer_0 *= 0
        for word in review.split(" "):
            self.layer_0[0][self.word2index[word]] = 1

    def get_target_for_label(self,label):
        if(label == 'POSITIVE'):
            return 1
        else:
            return 0
        
    def train(self, training_reviews_raw, training_labels):

        training_reviews = list()
        for review in training_reviews_raw:
            indices = set()
            for word in review.split(" "):
                if(word in self.word2index.keys()):
                    indices.add(self.word2index[word])
            training_reviews.append(list(indices))

        layer_1 = np.zeros_like(self.weights_0_1[0])

        start = time.time()
        correct_so_far = 0
        total_pred = 0.5
        for i in range(len(training_reviews_raw)):
            review_indices = training_reviews[i]
            label = training_labels[i]

            layer_1 *= 0
            for index in review_indices:
                layer_1 += self.weights_0_1[index]
            layer_1 = layer_1 / float(len(review_indices))
            layer_1 = layer_1.astype('int64') # round to nearest integer

            layer_2 = sigmoid(innerProd(layer_1,self.weights_1_2[0],M_onehot[len(layer_1) - 2][1],l) / float(scaling_factor))[0:2]

            if(label == 'POSITIVE'):
                layer_2_delta = layer_2 - (c_ones[len(layer_2) - 2] * scaling_factor)
            else:
                layer_2_delta = layer_2

            weights_1_2_trans = transpose(self.weights_1_2)
            layer_1_delta = mat_mul_forward(layer_2_delta,weights_1_2_trans,scaling_factor).astype('int64')

            self.weights_1_2 -= np.array(outer_product(layer_2_delta,layer_1))  * self.learning_rate

            for index in review_indices:
                self.weights_0_1[index] -= (layer_1_delta * self.learning_rate).astype('int64')

            # we're going to decrypt on the fly so we can watch what's happening
            total_pred += (s_decrypt(layer_2)[0] / scaling_factor)
            if((s_decrypt(layer_2)[0] / scaling_factor) >= (total_pred / float(i+2)) and label == 'POSITIVE'):
                correct_so_far += 1
            if((s_decrypt(layer_2)[0] / scaling_factor) < (total_pred / float(i+2)) and label == 'NEGATIVE'):
                correct_so_far += 1

            reviews_per_second = i / float(time.time() - start)

            sys.stdout.write("\rProgress:" + str(100 * i/float(len(training_reviews_raw)))[:4] + "% Speed(reviews/sec):" + str(reviews_per_second)[0:5] + " #Correct:" + str(correct_so_far) + " #Trained:" + str(i+1) + " Training Accuracy:" + str(correct_so_far * 100 / float(i+1))[:4] + "%")
            if(i % 100 == 0):
                print(i)

    
    def test(self, testing_reviews, testing_labels):
        
        correct = 0
        
        start = time.time()
        
        for i in range(len(testing_reviews)):
            pred = self.run(testing_reviews[i])
            if(pred == testing_labels[i]):
                correct += 1
            
            reviews_per_second = i / float(time.time() - start)
            
            sys.stdout.write("\rProgress:" + str(100 * i/float(len(testing_reviews)))[:4] \
                             + "% Speed(reviews/sec):" + str(reviews_per_second)[0:5] \
                            + "% #Correct:" + str(correct) + " #Tested:" + str(i+1) + " Testing Accuracy:" + str(correct * 100 / float(i+1))[:4] + "%")
    
    def run(self, review):
        
        # Input Layer


        # Hidden layer
        self.layer_1 *= 0
        unique_indices = set()
        for word in review.lower().split(" "):
            if word in self.word2index.keys():
                unique_indices.add(self.word2index[word])
        for index in unique_indices:
            self.layer_1 += self.weights_0_1[index]
        
        # Output layer
        layer_2 = self.sigmoid(self.layer_1.dot(self.weights_1_2))
        
        if(layer_2[0] >= 0.5):
            return "POSITIVE"
        else:
            return "NEGATIVE"


In [31]:
mlp = SentimentNetwork(reviews[:-1000],labels[:-1000],min_count=20,polarity_cutoff=0.15,learning_rate=0.01)

Pre-processing data...
Initializing Weights...
Encrypting Weights...
Encrypting Weights from Layer 0 to Layer 1:100.% done
Encrypting Weights from Layer 1 to Layer 2:100.% done

In [32]:
saved_weights = (copy.deepcopy(mlp.weights_0_1),copy.deepcopy(mlp.weights_1_2))

In [33]:
mlp.weights_0_1 = copy.deepcopy(saved_weights[0])
mlp.weights_1_2 = copy.deepcopy(saved_weights[1])
mlp.learning_rate = 0.0125
mlp.train(reviews[0:-1000],labels[:-1000])

Progress:0.0% Speed(reviews/sec):0.0 #Correct:0 #Trained:1 Training Accuracy:0.0%0
Progress:0.41% Speed(reviews/sec):1.800 #Correct:0 #Trained:101 Training Accuracy:0.0%100
Progress:0.83% Speed(reviews/sec):1.783 #Correct:0 #Trained:201 Training Accuracy:0.0%200
Progress:1.25% Speed(reviews/sec):2.043 #Correct:0 #Trained:301 Training Accuracy:0.0%300
Progress:1.66% Speed(reviews/sec):2.310 #Correct:0 #Trained:401 Training Accuracy:0.0%400
Progress:2.08% Speed(reviews/sec):2.516 #Correct:0 #Trained:501 Training Accuracy:0.0%500
Progress:2.5% Speed(reviews/sec):2.664 #Correct:0 #Trained:601 Training Accuracy:0.0%600
Progress:2.91% Speed(reviews/sec):2.778 #Correct:0 #Trained:701 Training Accuracy:0.0%700
Progress:3.33% Speed(reviews/sec):2.866 #Correct:0 #Trained:801 Training Accuracy:0.0%800
Progress:3.75% Speed(reviews/sec):2.868 #Correct:0 #Trained:901 Training Accuracy:0.0%900
Progress:4.16% Speed(reviews/sec):2.826 #Correct:0 #Trained:1001 Training Accuracy:0.0%1000
Progress:4.58% S

KeyboardInterrupt: 